include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh

```diff
@@ -113,20 +120,76 @@ class open_addressing_ref_impl {
   /**
    * @brief Constructs open_addressing_ref_impl.
    *
+   * @param empty_key_sentinel The reserved key value for empty slots
    * @param empty_slot_sentinel Sentinel indicating an empty slot
+   * @param predicate Key equality binary callable
    * @param probing_scheme Probing scheme
    * @param storage_ref Non-owning ref of slot storage
    */
   __host__ __device__ explicit constexpr open_addressing_ref_impl(
+    cuco::empty_key<key_type> empty_key_sentinel,
     value_type empty_slot_sentinel,
+    key_equal const& predicate,
     probing_scheme_type const& probing_scheme,
     storage_ref_type storage_ref) noexcept
     : empty_slot_sentinel_{empty_slot_sentinel},
+      predicate_{empty_key_sentinel, predicate},
       probing_scheme_{probing_scheme},
       storage_ref_{storage_ref}
   {
   }
 
+  /**
+   * @brief Gets the sentinel value used to represent an empty key slot.
+   *
+   * @return The sentinel value used to represent an empty key slot
+   */
+  [[nodiscard]] __host__ __device__ constexpr auto empty_key_sentinel() const noexcept
+  {
+    return this->predicate_.empty_sentinel_;
+  }
+
+  /**
+   * @brief Gets the sentinel used to represent an empty slot.
+   *
+   * @return The sentinel value used to represent an empty slot
+   */
+  [[nodiscard]] __device__ constexpr value_type empty_slot_sentinel() const noexcept
+  {
+    return empty_slot_sentinel_;
+  }
+
+  /**
+   * @brief Returns the function that compares keys for equality.
+   *
+   * @return The key equality predicate
+   */
+  [[nodiscard]] __host__ __device__ constexpr detail::equal_wrapper<key_type, key_equal> key_eq()
```

Comment:
```suggestion
  [[nodiscard]] __host__ __device__ constexpr detail::equal_wrapper<key_type, key_equal> const& key_eq()
```
Similar to `probing_scheme`, probably returns by `const&` instead of by value.

Reply:
Should we make `empty_key_sentinel()` and `empty_slot_sentinel()` return `const&`, too? If we add support for larger types, this might become beneficial. I'll benchmark this one too.

Reply:
Equal wrapper and probing scheme (probably sentinel slot also) would be always returned by ref since it's used internally only and we know copying is not necessary.

As for key sentinel, depending on the use case, if it's a public API in data structures, returning by value aligns with std behavior and the performance wouldn't matter in that case. If it's in the implementation details (like here in the OA ref base class), maybe returning by ref is preferrable?

Reply:
It's public API for `static_map` and `static_set`, so those function need to comply with the STL. The OA impl could return a reference, since it is only used internally.

Reply:
The benchmark results are a bit inconclusive:

Reply:
If we addidionally make the `empty_*_sentinel()` functions return a ref we get:
copy vs. everything ref:
# static_set_insert_unique_occupancy

## [0] NVIDIA RTX A6000

|  Key  |  Distribution  |  Occupancy  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |        Diff |   %Diff |  Status  |
|-------|----------------|-------------|------------|-------------|------------|-------------|-------------|---------|----------|
|  I32  |     UNIQUE     |     0.1     |  43.455 ms |       2.83% |  43.161 ms |       3.15% | -294.272 us |  -0.68% |   PASS   |
|  I32  |     UNIQUE     |     0.2     |  43.326 ms |       2.19% |  43.069 ms |       3.12% | -257.351 us |  -0.59% |   PASS   |
|  I32  |     UNIQUE     |     0.3     |  42.708 ms |       3.17% |  42.461 ms |       3.58% | -247.056 us |  -0.58% |   PASS   |
|  I32  |     UNIQUE     |     0.4     |  43.404 ms |       2.42% |  43.318 ms |       2.82% |  -86.169 us |  -0.20% |   PASS   |
|  I32  |     UNIQUE     |     0.5     |  42.228 ms |       1.75% |  42.588 ms |       2.92% |  359.549 us |   0.85% |   PASS   |
|  I32  |     UNIQUE     |     0.6     |  42.242 ms |       1.31% |  43.028 ms |       1.01% |  786.802 us |   1.86% |   FAIL   |
|  I32  |     UNIQUE     |     0.7     |  42.135 ms |       0.73% |  42.703 ms |       0.04% |  567.946 us |   1.35% |   FAIL   |
|  I32  |     UNIQUE     |     0.8     |  43.044 ms |       1.08% |  43.570 ms |       1.12% |  526.099 us |   1.22% |   FAIL   |
|  I32  |     UNIQUE     |     0.9     |  45.407 ms |       0.08% |  45.788 ms |       0.06% |  381.789 us |   0.84% |   FAIL   |
|  I64  |     UNIQUE     |     0.1     |  45.105 ms |       3.22% |  45.367 ms |       2.33% |  261.648 us |   0.58% |   PASS   |
|  I64  |     UNIQUE     |     0.2     |  44.373 ms |       2.86% |  44.203 ms |       3.42% | -170.016 us |  -0.38% |   PASS   |
|  I64  |     UNIQUE     |     0.3     |  44.851 ms |       1.87% |  44.338 ms |       2.47% | -513.147 us |  -1.14% |   PASS   |
|  I64  |     UNIQUE     |     0.4     |  44.468 ms |       2.81% |  45.275 ms |       2.21% |  807.341 us |   1.82% |   PASS   |
|  I64  |     UNIQUE     |     0.5     |  44.881 ms |       2.66% |  45.939 ms |       2.23% |    1.058 ms |   2.36% |   FAIL   |
|  I64  |     UNIQUE     |     0.6     |  45.511 ms |       1.78% |  45.582 ms |       1.56% |   70.513 us |   0.15% |   PASS   |
|  I64  |     UNIQUE     |     0.7     |  46.298 ms |       2.09% |  46.162 ms |       3.14% | -136.134 us |  -0.29% |   PASS   |
|  I64  |     UNIQUE     |     0.8     |  46.323 ms |       1.12% |  46.207 ms |       1.47% | -115.657 us |  -0.25% |   PASS   |
|  I64  |     UNIQUE     |     0.9     |  48.824 ms |       0.14% |  48.804 ms |       0.83% |  -20.512 us |  -0.04% |   PASS   |